Kaituo forecasting (#10133)
opensearch-project/documentation-website: _observing-your-data/forecast/security.md

        - 'indices:admin/mappings/fields/get*'
        - 'indices:admin/mappings/get'
        - 'indices:admin/resolve/index'
        - 'indices:data/read*'
        - 'indices:data/read/field_caps*'
        - 'indices:data/read/search'
        - 'indices:data/write*'
        - 'indices_monitor'
```

These roles do not include default `index_permissions` for specific source or result indices. This is intentional, allowing you to add your own patterns based on your data access requirements.

### Data-source read role

Each forecaster uses the creating user’s credentials to query the source index. To enable this, you must grant that user read permissions on your own data index.

The following example creates a minimal role that allows read access to the `network-metrics` index:

```json
PUT _plugins/_security/api/roles/data_source_read
{
  "index_permissions": [{
    "index_patterns": ["network-metrics"],
    "allowed_actions": ["read"]
  }]
}
```

You can modify the `index_patterns` to match your actual data source.

### Result‑read role

The result-read role allows users to view forecast results and configure Alerting monitors that query those results.

> The following example defines a role that grants read access to all indices matching the `opensearch-forecast-result*` pattern:

Author: ```suggestion
The following example defines a role that grants read access to all indexes matching the `opensearch-forecast-result*` pattern:
```

          "allowed_actions":["search"]
        }]
      }'

# Map the role to forecastuser
curl -XPUT -k -u 'admin:<custom-admin-password>' \
  'https://localhost:9200/_plugins/_security/api/rolesmapping/data_source_read' \
  -H 'Content-Type: application/json' \
  -d '{"users":["forecastuser"]}'
```
{% include copy-curl.html %}

In the remote cluster, define the same role and map it to the same user to ensure permissions are mirrored across clusters, as shown in the following:

```
# Create the identical role
curl -XPUT -k -u 'admin:<custom-admin-password>' \
  'https://localhost:9250/_plugins/_security/api/roles/data_source_read' \
  -H 'Content-Type: application/json' \
  -d '{
        "index_permissions":[{
          "index_patterns":["network-requests"],
          "allowed_actions":["search"]
        }]
      }'

# Map the role to the same user
curl -XPUT -k -u 'admin:<custom-admin-password>' \
  'https://localhost:9250/_plugins/_security/api/rolesmapping/data_source_read' \
  -H 'Content-Type: application/json' \
  -d '{"users":["forecastuser"]}'
```
{% include copy-curl.html %}

### Register the remote cluster with the local cluster

> Register the remote cluster on the local cluster using a seed node, under the under `cluster.remote.<alias>.seeds` setting,  In OpenSearch, this is called adding a `follower` cluster.

Author: ```suggestion
Register the remote cluster on the local cluster using a seed node, under the `cluster.remote.<alias>.seeds` setting,  In OpenSearch, this is called adding a `follower` cluster.
```

Assuming that the remote cluster is listening on transport port `9350`, run the following command on the local cluster:

```
curl -X PUT "https://localhost:9200/_cluster/settings" \
  -H "Content-Type: application/json" \
  -u "admin:<custom-admin-password>" \
  -d '{
    "persistent": {
      "cluster.remote": {
        "follower": {
          "seeds": [ "127.0.0.1:9350" ]
        }
      }
    }
  }'
```
{% include copy-curl.html %}

- Replace `127.0.0.1` with the remote node’s transport-layer IP if it's on a different host.

> - The alias `follower` can be any name you choose and will be used when referencing remote indices or configuring cross-cluster replication.

Author: ```suggestion
- The alias `follower` can be any name you choose and will be used when referencing remote indexes or configuring cross-cluster replication.
```

If the custom index already exists, the Forecasting plugin checks that the index mapping matches the expected forecast result structure. To ensure compatibility, the index must conform to the schema defined in the [`forecast-results.json`](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/resources/mappings/forecast-results.json) file.

When a user creates a forecaster—either in OpenSearch Dashboards or by calling the Forecasting API—the system verifies that the user has the following index-level permissions on the custom index:

- `indices:admin/create` – Required to create and roll over the custom result index.
- `indices:admin/aliases` – Required to create and manage the index alias.
- `indices:data/write/index` – Required to write forecast results into the index (single-stream forecasters).
- `indices:data/read/search` – Required to search the custom index when displaying forecast results.
- `indices:data/write/delete` – Required to delete older forecast results and manage disk usage.
- `indices:data/write/bulk*` – Required because the plugin writes results using the Bulk API.

## Next step

> For additional guidance on Transport Layer Security (TLS), authentication backends, tenant isolation, and audit logging, refer to the [Security plugin documentation]({{site.url}}{{site.baseurl}}/security).

Author: ```suggestion
For additional guidance on Transport Layer Security (TLS), authentication backends, tenant isolation, and audit logging, refer to the [Security plugin documentation]({{site.url}}{{site.baseurl}}/security/).
```

## Required roles

A forecasting user needs three types of privileges, based on the following responsibilities:

- *Managing the forecasting job*
- *Reading the source data*
- *Accessing the forecast results*

These responsibilities correspond to three distinct security layers, as shown in the following table:

| Layer | What it controls | Typical role |
|-------|------------------|--------------|
| **Forecaster control** | Permissions to create, edit, start, stop, delete, or view a forecaster’s configuration. | `forecast_full_access` <br>*(manage lifecycle)*<br>or<br>`forecast_read_access` <br>*(view only)* |
| **Data-source read** | Grants the forecaster permission to query the raw metrics index it uses for training and prediction. | Custom role, such as `data_source_read` |
| **Result read** | Grants users and Alerting monitors access to documents in `opensearch-forecast-result*`. | Custom role, such as `forecast_result_read` |

The built-in roles `forecast_full_access` and `forecast_read_access` apply only to Forecasting APIs. They do **not** include permissions for source or result indexes—those must be granted separately.
{: .note}

> ### Forecaster‑control roles

Reviewer: Is this intentionally hyphenated?

> ### Forecaster‑control roles

Author: ```suggestion
### Forecaster control roles
```

      allowed_actions:
        - 'indices:admin/aliases/get'
        - 'indices:admin/mapping/get'
        - 'indices:admin/mapping/put'
        - 'indices:admin/mappings/fields/get*'
        - 'indices:admin/mappings/get'
        - 'indices:admin/resolve/index'
        - 'indices:data/read*'
        - 'indices:data/read/field_caps*'
        - 'indices:data/read/search'
        - 'indices:data/write*'
        - 'indices_monitor'
```
{% include copy.html %}

These roles do not include default `index_permissions` for specific source or result indexes. This is intentional, allowing you to add your own patterns based on your data access requirements.

### Data-source read role

Each forecaster uses the creating user’s credentials to query the source index. To enable this, you must grant that user read permissions on your own data index.

The following example creates a minimal role that allows read access to the `network-metrics` index:

```json
PUT _plugins/_security/api/roles/data_source_read
{
  "index_permissions": [{
    "index_patterns": ["network-metrics"],
    "allowed_actions": ["read"]
  }]
}
```

You can modify the `index_patterns` to match your actual data source.

> ### Result‑read role

Reviewer: Is this intentionally hyphenated? Should the name of the role be formatted differently (in code font)?

> ### Result‑read role

Author: ```suggestion
### `Result‑read` role
```

> The result-read role allows users to view forecast results and configure Alerting monitors that query those results.

Reviewer: Is this intentionally hyphenated? Should the name of the role be formatted differently (in code font)?

> The result-read role allows users to view forecast results and configure Alerting monitors that query those results.

Author: ```suggestion
The `forecast_result_read` role allows users to view forecast results and configure Alerting monitors that query those results.
```

> The following example defines a role that grants read access to all indexes matching the `opensearch-forecast-result*` pattern:

Reviewer: ```suggestion
The following example request defines a role that grants read access to all indexes matching the `opensearch-forecast-result*` pattern:
```

```json
PUT _plugins/_security/api/roles/forecast_result_read
{
  "index_permissions": [{
    "index_patterns": ["opensearch-forecast-result*"],
    "allowed_actions": ["read"]
  }]
}
```

If you need to isolate result data between teams, you can enhance this role using document-level security (DLS) with a backend role filter, as shown in the next section

### Example security role configuration

The following example creates a `devOpsEngineer` user and assigns all three required roles for forecasting:

```json
PUT _plugins/_security/api/internalusers/devOpsEngineer
{
  "password": "DevOps2024!",
  "opendistro_security_roles": [
    "forecast_full_access",
    "data_source_read",
    "forecast_result_read"
  ]
}
```

> 

Reviewer: Should these examples include copy/copy curl buttons?

This configuration enables the following:

- `devOpsEngineer` can manage forecasters (`forecast_full_access`).
- Forecasters can query the source index successfully (`data_source_read`).
- The user and any configured monitors can read forecast results (`forecast_result_read`).

To grant read-only access to forecaster configurations, replace `forecast_full_access` with `forecast_read_access`.

---

## (Advanced) Limit access by backend role

> You can use backend roles to enforce **team-specific isolation** in Forecasting. This pattern allows different teams to operate forecasters independently while keeping configurations and results separated.

Reviewer: ```suggestion
You can use backend roles to enforce **team-specific isolation**. This pattern allows different teams to operate forecasters independently while separating configurations and results.
```

The model includes three layers:

1. **Configuration isolation** – Forecasting APIs are restricted to users with a matching backend role.

> 2. **Result isolation** – Document-level security (DLS) limits access to forecast results in `opensearch-forecast-result*`.

Reviewer: ```suggestion
2. **Result isolation** – DLS limits access to forecast results in `opensearch-forecast-result*`.
```

The following sections explain how to configure each layer.

### Assign backend roles to users

In most environments, backend roles are assigned through LDAP or SAML. However, if you are using the internal user database, you can set them manually as shown in the following example:

```json
# Analyst
PUT _plugins/_security/api/internalusers/alice
{
  "password": "alice",
  "backend_roles": ["analyst"]
}

# HR staff
PUT _plugins/_security/api/internalusers/bob
{
  "password": "bob",
  "backend_roles": ["human-resources"]
}
```

These backend roles can then be used to control access to forecasters and forecast results on a per-team basis.

### Enable backend-role filtering for configuration access

To isolate forecaster configurations by team, enable backend-role filtering at the cluster level:

```bash
PUT _cluster/settings
{
  "persistent": {
    "plugins.forecast.filter_by_backend_roles": true
  }
}
```

When this setting is enabled, OpenSearch records the creator’s backend roles in each forecaster document. Only users with a matching backend role can view, edit, or delete that forecaster.

### Create a result‑access role per team

> Forecast results are stored in shared indexes, so use document-level security (DLS) to restrict access by backend role.

Reviewer: ```suggestion
Forecast results are stored in shared indexes, so use DLS to restrict access by backend role.
```

> The following example creates a role that allows users with the `analyst` backend role to read and write only their team’s forecast results:

Reviewer: ```suggestion
The following example request creates a role that allows users with the `analyst` backend role to read and write only their team's forecast results:
```

```json
PUT _plugins/_security/api/roles/forecast_analyst_result_access
{
  "index_permissions": [{
    "index_patterns": ["opensearch-forecast-result*"],
    "dls": """
    {
      "bool": {
        "filter": [{
          "nested": {
            "path": "user",
            "query": {
              "term": {
                "user.backend_roles.keyword": "analyst"
              }
            },
            "score_mode": "none"
          }
        }]
      }
    }""",
    "allowed_actions": ["read","write"]
  }]
}
```

To isolate results for another team, such as `human-resources`, create a separate role (for example, `forecast_human_resources_result_access`) and update the term value to match the appropriate backend role.

> ### Define data-source read access

Reviewer: Is this intentionally hyphenated?

The `data_source_read` role is defined the same way as in earlier examples. It grants minimal read access to the metrics index that each forecaster uses for training and prediction.

> You can reuse this role across teams, or create separate versions if you need per-index restrictions.

Reviewer: ```suggestion
You can reuse this role across teams or create separate versions if you need per-index restrictions.
```

### Map a user to three roles

> The following example maps the user `alice` to all three required roles—forecaster control, result access, and source index read—using the `analyst` backend role:

Reviewer: This is getting confusing. Should we use the actual names of the roles here (in code font)?

> The following example maps the user `alice` to all three required roles—forecaster control, result access, and source index read—using the `analyst` backend role:

Author: ```suggestion
The following example maps the user `alice` to all three required roles—`forecaster-control`, `result-access`, and `data_source_read`—using the `analyst` backend role:
```

```json
PUT _plugins/_security/api/internalusers/alice
{
  "password": "alice",
  "backend_roles": ["analyst"],
  "opendistro_security_roles": [
    "forecast_full_access",
    "forecast_analyst_result_access",
    "data_source_read"
  ]
}
```

With this configuration, Alice can:

- Create, start, stop, and delete only forecasters tagged with the analyst backend role.
- View only forecast results tagged for analyst.

> - Read the network-metrics index as the source for her forecasters.

Reviewer: ```suggestion
- Read the `network-metrics` index as the source for her forecasters.
```

To configure a second user such as `bob` from the HR team, use a parallel setup with the human-resources backend role and forecast_human_resources_result_access.

### What if users lack backend roles?

> If a user has the `forecast_read_access` role but no backend roles, they cannot view any forecasters. Backend-role filtering enforces strict matching and prevents access to configurations that do not align with the user’s roles.

Reviewer: ```suggestion
If a user has the `forecast_read_access` role but no backend roles, they cannot view any forecasters. Backend-role filtering enforces strict matching and prevents access to configurations that do not align with the user's roles.
```

  -H 'Content-Type: application/json' \
  -d '{
        "index_permissions":[{
          "index_patterns":["network-requests"],
          "allowed_actions":["search"]
        }]
      }'

# Map the role to the same user
curl -XPUT -k -u 'admin:<custom-admin-password>' \
  'https://localhost:9250/_plugins/_security/api/rolesmapping/data_source_read' \
  -H 'Content-Type: application/json' \
  -d '{"users":["forecastuser"]}'
```
{% include copy-curl.html %}

### Register the remote cluster with the local cluster

Register the remote cluster with the local cluster using a seed node under the `cluster.remote.<alias>.seeds` setting. In OpenSearch, this is called adding a `follower` cluster.

Assuming that the remote cluster is listening on transport port `9350`, run the following command in the local cluster:

```
curl -X PUT "https://localhost:9200/_cluster/settings" \
  -H "Content-Type: application/json" \
  -u "admin:<custom-admin-password>" \
  -d '{
    "persistent": {
      "cluster.remote": {
        "follower": {
          "seeds": [ "127.0.0.1:9350" ]
        }
      }
    }
  }'
```
{% include copy-curl.html %}

> - Replace `127.0.0.1` with the remote node's transport layer IP if it's on a different host.

Author: ```suggestion
- Replace `127.0.0.1` with the remote node's transport layer IP if it's located on a different host.
```

    - 'cluster:admin/plugin/forecast/tasks/search'
  index_permissions:
    - index_patterns:
        - 'opensearch-forecast-result*'
      allowed_actions:
        - 'indices:admin/mappings/fields/get*'
        - 'indices:admin/resolve/index'
        - 'indices:data/read*'

forecast_full_access:
  reserved: true
  cluster_permissions:
    - 'cluster:admin/plugin/forecast/*'
    - 'cluster:admin/settings/update'
    - 'cluster_monitor'
  index_permissions:
    - index_patterns:
        - '*'
      allowed_actions:
        - 'indices:admin/aliases/get'
        - 'indices:admin/mapping/get'
        - 'indices:admin/mapping/put'
        - 'indices:admin/mappings/fields/get*'
        - 'indices:admin/mappings/get'
        - 'indices:admin/resolve/index'
        - 'indices:data/read*'
        - 'indices:data/read/field_caps*'
        - 'indices:data/read/search'
        - 'indices:data/write*'
        - 'indices_monitor'
```
{% include copy.html %}

These roles do not include default `index_permissions` for specific source or result indexes. This is intentional, allowing you to add your own patterns based on your data access requirements.

> ### Data-source read role

Author: ```suggestion
### Data source `read` role
```
